File: r-mos-frontend/src/api/training.ts
import { apiClient } from './client'

export interface SkillProfileResponse {
  user_id: number
  overall_level: number
  total_sessions: number
  total_duration: number
  last_trained_at?: string | null
  score_safety?: number | null
  score_procedure?: number | null
  score_precision?: number | null
  score_efficiency?: number | null
  score_tools?: number | null
  cert_l1_passed: boolean
  cert_l2_passed: boolean
  cert_l3_eligible: boolean
}

export interface WeakStepResponse {
  step_id: string
  sop_id?: string | null
  fail_count: number
  last_failed_at?: string | null
  fail_tags?: string[] | null
  is_resolved: boolean
}

export interface SessionResponse {
  session_id: string
  project_id: string
  user_id: number
  status: string
  current_step: number
  score?: number | null
  total_duration: number
  submit_type?: string | null
  started_at: string
  paused_at?: string | null
  submitted_at?: string | null
  project_snapshot?: {
    title?: string
    summary?: string
    robot_model?: string
    task_summary?: string
    focus_prompt?: string
    seed_messages?: Array<{
      id?: string
      role?: 'assistant' | 'teacher' | 'user'
      content?: string
      created_at?: string
    }>
    steps?: Array<{
      id: string
      step_index: number
      title: string
      instruction: string
      evidence_hint?: string
      model_targets?: string[]
      tools?: Array<{
        id: string
        name: string
        spec?: string
        is_critical?: boolean
        recommendation?: string
      }>
    }>
  } | null
}

export interface SessionDetailResponse {
  session: SessionResponse
  steps: Array<{
    record_id: string
    session_id: string
    step_id: string
    step_index: number
    status: string
    attempt_count: number
    duration_sec?: number | null
    tools_confirmed?: Array<{
      tool_id: string
      status: 'PENDING' | 'CONFIRMED' | 'ANOMALY'
    }> | null
    evidence?: {
      bundle_id?: string | null
      note?: string | null
    } | null
    verdict_result?: {
      result: 'PASS' | 'FAIL'
      summary: string
      details?: string
      missing_critical_tools?: string[]
      anomaly_tools?: string[]
      evidence_bundle_id?: string | null
    } | null
  }>
}

export interface TrainingWorkbenchDraftPayload {
  robotModel: string
  taskSummary: string
  focusPrompt: string
}

export interface TrainingWorkbenchDraftResponse {
  project: {
    sessionId: string
    projectId: string
    title: string
    progressPercent: number
  }
  steps: Array<{
    id: string
    stepIndex: number
    title: string
    status: 'pending' | 'active' | 'passed' | 'failed'
    instruction: string
    evidenceHint?: string
    modelTargets?: string[]
    tools: Array<{
      id: string
      name: string
      spec?: string
      isCritical?: boolean
      recommendation?: string
    }>
  }>
  messages: Array<{
    id: string
    role: 'assistant' | 'teacher' | 'user'
    content: string
    createdAt: string
  }>
}

export interface TrainingWorkbenchEvidenceUploadResponse {
  evidenceBundleId: string
  filename: string
  contentUri: string
  humanSummary?: string
}

export interface TrainingWorkbenchToolConfirmPayload {
  toolId: string
  status: 'PENDING' | 'CONFIRMED' | 'ANOMALY'
}

export interface TrainingWorkbenchStepSubmitPayload {
  stepIndex: number
  note: string
  evidenceBundleId?: string | null
  toolsConfirmed: TrainingWorkbenchToolConfirmPayload[]
}

export interface TrainingWorkbenchStepSubmitResponse {
  recordId: string
  status: 'pass' | 'fail'
  evidenceBundleId?: string | null
  nextStepId?: string | null
  sessionSubmitted: boolean
  feedback?: {
    submissionId?: string
    score?: number | null
  } | null
  verdict: {
    result: 'PASS' | 'FAIL'
    summary: string
    details?: string
    missingCriticalTools?: string[]
    anomalyTools?: string[]
    evidenceBundleId?: string | null
  }
}

export interface TrainingWorkbenchAssistantMessage {
  id: string
  role: 'assistant' | 'teacher' | 'user'
  content: string
  createdAt: string
}

export async function getStudentProfile(userId: number) {
  const response = await apiClient.get<SkillProfileResponse>(`/students/${userId}/profile`)
  return response.data
}

export async function getWeakSteps(userId: number) {
  const response = await apiClient.get<WeakStepResponse[]>(`/students/${userId}/weak-steps`)
  return response.data
}

export async function getTrainingSessions(userId: number) {
  const response = await apiClient.get<SessionResponse[]>(`/training/users/${userId}/sessions`)
  return response.data
}

export async function getActiveTrainingSession(userId: number) {
  const response = await apiClient.get<SessionResponse>(`/training/users/${userId}/active-session`)
  return response.data
}

export async function getTrainingSessionDetail(sessionId: string) {
  const response = await apiClient.get<SessionDetailResponse>(`/training/sessions/${sessionId}/detail`)
  return response.data
}

export async function generateTrainingWorkbenchDraft(
  payload: TrainingWorkbenchDraftPayload,
): Promise<TrainingWorkbenchDraftResponse> {
  const response = await apiClient.post('/training/workbench/draft', {
    robot_model: payload.robotModel,
    task_summary: payload.taskSummary,
    focus_prompt: payload.focusPrompt,
  }, {
    timeout: 90000,
  })

  return {
    project: {
      sessionId: response.data.project.session_id,
      projectId: response.data.project.project_id,
      title: response.data.project.title,
      progressPercent: response.data.project.progress_percent,
    },
    steps: response.data.steps.map((step: Record<string, unknown>) => ({
      id: String(step.id),
      stepIndex: Number(step.step_index ?? 0),
      title: String(step.title),
      status: step.status as 'pending' | 'active' | 'passed' | 'failed',
      instruction: String(step.instruction),
      evidenceHint: typeof step.evidence_hint === 'string' ? step.evidence_hint : undefined,
      modelTargets: Array.isArray(step.model_targets)
        ? step.model_targets
            .filter((value): value is string => typeof value === 'string')
        : [],
      tools: Array.isArray(step.tools)
        ? step.tools.map((tool) => {
            const toolRecord = tool as Record<string, unknown>
            return {
              id: String(toolRecord.id),
              name: String(toolRecord.name),
              spec: typeof toolRecord.spec === 'string' ? toolRecord.spec : undefined,
              isCritical: Boolean(toolRecord.is_critical),
              recommendation:
                typeof toolRecord.recommendation === 'string' ? toolRecord.recommendation : undefined,
            }
          })
        : [],
    })),
    messages: response.data.messages.map((message: Record<string, unknown>) => ({
      id: String(message.id),
      role: message.role as 'assistant' | 'teacher' | 'user',
      content: String(message.content),
      createdAt: String(message.created_at),
    })),
  }
}

export async function uploadTrainingWorkbenchEvidence(
  sessionId: string,
  stepId: string,
  note: string,
  file: File,
): Promise<TrainingWorkbenchEvidenceUploadResponse> {
  const formData = new FormData()
  formData.append('session_id', sessionId)
  formData.append('step_id', stepId)
  formData.append('note', note)
  formData.append('file', file)

  const response = await apiClient.post('/training/workbench/evidence', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
    timeout: 90000,
  })

  return {
    evidenceBundleId: String(response.data.evidence_bundle_id),
    filename: String(response.data.filename),
    contentUri: String(response.data.content_uri),
    humanSummary:
      typeof response.data.human_summary === 'string' ? response.data.human_summary : undefined,
  }
}

export async function submitTrainingWorkbenchStep(
  sessionId: string,
  stepId: string,
  payload: TrainingWorkbenchStepSubmitPayload,
): Promise<TrainingWorkbenchStepSubmitResponse> {
  const response = await apiClient.post(
    `/training/workbench/sessions/${sessionId}/steps/${stepId}/submit`,
    {
      step_index: payload.stepIndex,
      note: payload.note,
      evidence_bundle_id: payload.evidenceBundleId ?? null,
      tools_confirmed: payload.toolsConfirmed.map((tool) => ({
        tool_id: tool.toolId,
        status: tool.status,
      })),
    },
    {
      timeout: 90000,
    },
  )

  return {
    recordId: String(response.data.record_id),
    status: response.data.status as 'pass' | 'fail',
    evidenceBundleId:
      typeof response.data.evidence_bundle_id === 'string' ? response.data.evidence_bundle_id : null,
    nextStepId: typeof response.data.next_step_id === 'string' ? response.data.next_step_id : null,
    sessionSubmitted: Boolean(response.data.session_submitted),
    feedback: response.data.feedback
      ? {
          submissionId:
            typeof response.data.feedback.submission_id === 'string'
              ? response.data.feedback.submission_id
              : undefined,
          score:
            typeof response.data.feedback.score === 'number' ? response.data.feedback.score : null,
        }
      : null,
    verdict: {
      result: response.data.verdict.result as 'PASS' | 'FAIL',
      summary: String(response.data.verdict.summary),
      details:
        typeof response.data.verdict.details === 'string' ? response.data.verdict.details : undefined,
      missingCriticalTools: Array.isArray(response.data.verdict.missing_critical_tools)
        ? response.data.verdict.missing_critical_tools
            .filter((value: unknown): value is string => typeof value === 'string')
        : [],
      anomalyTools: Array.isArray(response.data.verdict.anomaly_tools)
        ? response.data.verdict.anomaly_tools
            .filter((value: unknown): value is string => typeof value === 'string')
        : [],
      evidenceBundleId:
        typeof response.data.verdict.evidence_bundle_id === 'string'
          ? response.data.verdict.evidence_bundle_id
          : undefined,
    },
  }
}

export async function askTrainingWorkbenchAssistant(payload: {
  sessionId: string
  stepId: string
  question: string
  messages: Array<{
    role: 'assistant' | 'teacher' | 'user'
    content: string
  }>
}): Promise<TrainingWorkbenchAssistantMessage> {
  const response = await apiClient.post(
    '/training/workbench/ask',
    {
      session_id: payload.sessionId,
      step_id: payload.stepId,
      question: payload.question,
      messages: payload.messages,
    },
    {
      timeout: 90000,
    },
  )

  return {
    id: String(response.data.id),
    role: response.data.role as 'assistant' | 'teacher' | 'user',
    content: String(response.data.content),
    createdAt: String(response.data.created_at),
  }
}

File: r-mos-backend/app/api/v1/endpoints/training.py
"""
UF-04, UF-06: Training API Endpoints
训练项目与会话管理接口
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional, List
import logging

from app.core.database import get_db
from app.models.audit_event import AuditEvent
from app.services.training.session_service import SessionService
from app.services.training.submission_service import SubmissionService
from app.services.training.feedback_generator import FeedbackGenerator, FeedbackRole
from app.services.identity.class_membership import ClassMembershipService
from app.services.memory.skill_profile_service import SkillProfileService
from app.models.training_submission import TrainingSubmission
from app.schemas.training_workbench import (
    SessionCreateRequest, SessionResponse, StepRecordResponse,
    StepUpdateRequest, SessionDetailResponse, SubmitSessionRequest,
    ForceSubmitSessionRequest, SubmitSessionResponse,
    SkillProfileResponse, WeakStepResponse, FeedbackResponse,
)
from app.api.v1.endpoints import training_workbench

logger = logging.getLogger(__name__)

router = APIRouter()

router.include_router(training_workbench.router)


# ============ UF-06: Session Routes ============

@router.post(
    "/training/sessions",
    response_model=SessionResponse,
    tags=["Training"]
)
async def create_session(
    request: SessionCreateRequest,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-1: 创建训练会话"""
    service = SessionService(db)
    session_id = await service.create_session(
        user_id=request.user_id,
        project_id=request.project_id,
        project_snapshot=request.project_snapshot,
        ab_group=request.ab_group,
    )

    # 获取创建后的会话
    session = await service.get_session(session_id)
    if not session:
        raise HTTPException(status_code=500, detail="Failed to create session")

    return SessionResponse(
        session_id=session.session_id,
        project_id=session.project_id,
        user_id=session.user_id,
        status=session.status,
        current_step=session.current_step,
        score=session.score,
        total_duration=session.total_duration,
        submit_type=session.submit_type,
        started_at=session.started_at,
        paused_at=session.paused_at,
        submitted_at=session.submitted_at,
        project_snapshot=session.project_snapshot,
    )


@router.get(
    "/training/sessions/{session_id}",
    response_model=SessionResponse,
    tags=["Training"]
)
async def get_session(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-3: 获取会话状态"""
    service = SessionService(db)
    session = await service.get_session(session_id)

    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    return SessionResponse(
        session_id=session.session_id,
        project_id=session.project_id,
        user_id=session.user_id,
        status=session.status,
        current_step=session.current_step,
        score=session.score,
        total_duration=session.total_duration,
        submit_type=session.submit_type,
        started_at=session.started_at,
        paused_at=session.paused_at,
        submitted_at=session.submitted_at,
        project_snapshot=session.project_snapshot,
    )


@router.get(
    "/training/sessions/{session_id}/detail",
    response_model=SessionDetailResponse,
    tags=["Training"]
)
async def get_session_detail(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-3: 获取会话详情（含步骤记录）"""
    service = SessionService(db)
    result = await service.get_session_with_steps(session_id)

    if not result:
        raise HTTPException(status_code=404, detail="Session not found")

    session = result["session"]
    steps = result["steps"]

    return SessionDetailResponse(
        session=SessionResponse(
            session_id=session.session_id,
            project_id=session.project_id,
            user_id=session.user_id,
            status=session.status,
            current_step=session.current_step,
            score=session.score,
            total_duration=session.total_duration,
            submit_type=session.submit_type,
            started_at=session.started_at,
            paused_at=session.paused_at,
            submitted_at=session.submitted_at,
            project_snapshot=session.project_snapshot,
        ),
        steps=[
            StepRecordResponse(
                record_id=s.record_id,
                session_id=s.session_id,
                step_id=s.step_id,
                step_index=s.step_index,
                status=s.status,
                attempt_count=s.attempt_count,
                duration_sec=s.duration_sec,
                tools_confirmed=s.tools_confirmed,
                evidence=s.evidence,
                verdict_result=s.verdict_result,
                started_at=s.started_at,
                completed_at=s.completed_at,
            )
            for s in steps
        ],
    )


@router.patch(
    "/training/sessions/{session_id}/pause",
    response_model=SessionResponse,
    tags=["Training"]
)
async def pause_session(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-1: 暂停会话"""
    service = SessionService(db)
    session = await service.pause(session_id)

    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    return SessionResponse(
        session_id=session.session_id,
        project_id=session.project_id,
        user_id=session.user_id,
        status=session.status,
        current_step=session.current_step,
        score=session.score,
        total_duration=session.total_duration,
        submit_type=session.submit_type,
        started_at=session.started_at,
        paused_at=session.paused_at,
        submitted_at=session.submitted_at,
        project_snapshot=session.project_snapshot,
    )


@router.patch(
    "/training/sessions/{session_id}/resume",
    response_model=SessionResponse,
    tags=["Training"]
)
async def resume_session(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-1: 恢复会话"""
    service = SessionService(db)
    session = await service.resume(session_id)

    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    return SessionResponse(
        session_id=session.session_id,
        project_id=session.project_id,
        user_id=session.user_id,
        status=session.status,
        current_step=session.current_step,
        score=session.score,
        total_duration=session.total_duration,
        submit_type=session.submit_type,
        started_at=session.started_at,
        paused_at=session.paused_at,
        submitted_at=session.submitted_at,
        project_snapshot=session.project_snapshot,
    )


@router.patch(
    "/training/sessions/{session_id}/abandon",
    tags=["Training"]
)
async def abandon_session(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-c-4: 放弃会话"""
    service = SessionService(db)
    success = await service.abandon(session_id)

    if not success:
        raise HTTPException(status_code=404, detail="Session not found")

    return {"message": "Session abandoned", "session_id": session_id}


@router.post(
    "/training/sessions/{session_id}/submit",
    response_model=SubmitSessionResponse,
    tags=["Training"]
)
async def submit_session(
    session_id: str,
    request: SubmitSessionRequest,
    db: AsyncSession = Depends(get_db)
):
    """UF-08: 手动提交训练（走 SubmissionService）"""
    service = SubmissionService(db)

    check_result = await service.check_submit_ready(session_id)
    if not check_result.can_submit:
        if check_result.message == "会话不存在":
            raise HTTPException(status_code=404, detail="Session not found")
        raise HTTPException(status_code=400, detail=check_result.message)

    if check_result.incomplete_steps and not request.confirm_incomplete:
        raise HTTPException(
            status_code=409,
            detail={
                "message": check_result.message,
                "incomplete_steps": check_result.incomplete_steps,
                "requires_confirmation": True,
            },
        )

    submission = await service.submit_manual(
        session_id=session_id,
        user_id=request.user_id,
        confirm_incomplete=request.confirm_incomplete,
    )

    if not submission:
        raise HTTPException(status_code=400, detail="Submit failed")

    return SubmitSessionResponse(
        submission_id=submission.submission_id,
        session_id=submission.session_id,
        user_id=submission.user_id,
        submit_type=submission.submit_type,
        score=submission.payload.get("score"),
    )


@router.post(
    "/training/sessions/{session_id}/force-submit",
    response_model=SubmitSessionResponse,
    tags=["Training"],
)
async def force_submit_session(
    session_id: str,
    request: ForceSubmitSessionRequest,
    db: AsyncSession = Depends(get_db),
):
    """教师强制提交训练（需 teacher 对 student 有班级管辖权）。"""
    session_service = SessionService(db)
    training_session = await session_service.get_session(session_id)
    if not training_session:
        raise HTTPException(status_code=404, detail="Session not found")

    membership_service = ClassMembershipService(db)
    has_scope = await membership_service.teacher_has_student_scope(
        teacher_id=request.teacher_id,
        student_id=training_session.user_id,
    )
    if not has_scope:
        raise HTTPException(status_code=403, detail="Teacher has no scope for this student")

    submission_service = SubmissionService(db)
    submission = await submission_service.submit_by_teacher(
        session_id=session_id,
        teacher_id=request.teacher_id,
    )
    if not submission:
        raise HTTPException(status_code=400, detail="Force submit failed")

    db.add(
        AuditEvent(
            actor_user_id=str(request.teacher_id),
            action="student_notified",
            resource_type="TrainingSession",
            resource_id=session_id,
            decision="allow",
            reason="teacher_force_submit",
            request_meta={"student_user_id": training_session.user_id},
        )
    )
    await db.commit()

    return SubmitSessionResponse(
        submission_id=submission.submission_id,
        session_id=submission.session_id,
        user_id=submission.user_id,
        submit_type=submission.submit_type,
        score=submission.payload.get("score"),
    )


@router.post(
    "/training/sessions/{session_id}/steps",
    tags=["Training"]
)
async def update_step(
    session_id: str,
    request: StepUpdateRequest,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-2: 更新步骤记录"""
    service = SessionService(db)

    record_id = await service.update_step(
        session_id=session_id,
        step_id=request.step_id,
        step_index=request.step_index,
        status=request.status,
        attempt_count=request.attempt_count,
        tools_confirmed=request.tools_confirmed,
        evidence=request.evidence,
        verdict_result=request.verdict_result,
        duration_sec=request.duration_sec,
    )

    return {"record_id": record_id, "session_id": session_id}


@router.get(
    "/training/sessions/{session_id}/steps",
    response_model=List[StepRecordResponse],
    tags=["Training"]
)
async def get_step_records(
    session_id: str,
    db: AsyncSession = Depends(get_db)
):
    """UF-06-b-3: 获取步骤记录列表"""
    service = SessionService(db)
    result = await service.get_session_with_steps(session_id)

    if not result:
        raise HTTPException(status_code=404, detail="Session not found")

    steps = result["steps"]
    return [
        StepRecordResponse(
            record_id=s.record_id,
            session_id=s.session_id,
            step_id=s.step_id,
            step_index=s.step_index,
            status=s.status,
            attempt_count=s.attempt_count,
            duration_sec=s.duration_sec,
            tools_confirmed=s.tools_confirmed,
            evidence=s.evidence,
            verdict_result=s.verdict_result,
            started_at=s.started_at,
            completed_at=s.completed_at,
        )
        for s in steps
    ]


@router.get(
    "/training/users/{user_id}/sessions",
    response_model=List[SessionResponse],
    tags=["Training"]
)
async def get_user_sessions(
    user_id: int,
    status: Optional[str] = None,
    limit: int = 10,
    db: AsyncSession = Depends(get_db)
):
    """获取用户会话列表"""
    service = SessionService(db)
    sessions = await service.get_user_sessions(user_id, status, limit)

    return [
        SessionResponse(
            session_id=s.session_id,
            project_id=s.project_id,
            user_id=s.user_id,
            status=s.status,
            current_step=s.current_step,
            score=s.score,
            total_duration=s.total_duration,
            submit_type=s.submit_type,
            started_at=s.started_at,
            paused_at=s.paused_at,
            submitted_at=s.submitted_at,
            project_snapshot=s.project_snapshot,
        )
        for s in sessions
    ]


@router.get(
    "/training/users/{user_id}/active-session",
    response_model=SessionResponse,
    tags=["Training"]
)
async def get_active_session(
    user_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取用户当前活跃会话（用于断点续训）"""
    service = SessionService(db)
    session = await service.get_user_active_session(user_id)

    if not session:
        raise HTTPException(status_code=404, detail="No active session found")

    return SessionResponse(
        session_id=session.session_id,
        project_id=session.project_id,
        user_id=session.user_id,
        status=session.status,
        current_step=session.current_step,
        score=session.score,
        total_duration=session.total_duration,
        submit_type=session.submit_type,
        started_at=session.started_at,
        paused_at=session.paused_at,
        submitted_at=session.submitted_at,
        project_snapshot=session.project_snapshot,
    )


@router.get(
    "/training/feedback/{session_id}",
    response_model=FeedbackResponse,
    tags=["Training"]
)
async def get_training_feedback(
    session_id: str,
    role: str = Query(default="student", pattern="^(student|teacher)$"),
    db: AsyncSession = Depends(get_db)
):
    """UF-09: 获取训练反馈报告"""
    result = await db.execute(
        select(TrainingSubmission)
        .where(TrainingSubmission.session_id == session_id)
        .order_by(TrainingSubmission.submitted_at.desc())
    )
    submission = result.scalars().first()
    if not submission:
        raise HTTPException(status_code=404, detail="Submission not found")

    if not submission.feedback:
        feedback_generator = FeedbackGenerator(db)
        feedback = await feedback_generator.generate(
            submission_id=submission.submission_id,
            role=FeedbackRole.TEACHER if role == "teacher" else FeedbackRole.STUDENT,
        )
        payload = {
            "overall_score": feedback.overall_score,
            "score_breakdown": feedback.score_breakdown,
            "step_analyses": [
                {
                    "step_id": s.step_id,
                    "step_index": s.step_index,
                    "status": s.status,
                    "attempt_count": s.attempt_count,
                    "analysis": s.analysis,
                    "suggestions": s.suggestions,
                    "ref_ids": s.ref_ids,
                }
                for s in feedback.step_analyses
            ],
            "suggestions": feedback.suggestions,
            "next_learning_plan": feedback.next_learning_plan,
            "teaching_diagnosis": feedback.teaching_diagnosis,
            "ranking_percentile": feedback.ranking_percentile,
            "hint_level_suggestion": feedback.hint_level_suggestion,
        }
    else:
        payload = dict(submission.feedback)

    return FeedbackResponse(
        session_id=session_id,
        submission_id=submission.submission_id,
        overall_score=float(payload.get("overall_score", 0)),
        score_breakdown=payload.get("score_breakdown", {}),
        step_analyses=payload.get("step_analyses", []),
        suggestions=payload.get("suggestions", []),
        next_learning_plan=payload.get("next_learning_plan", ""),
        teaching_diagnosis=payload.get("teaching_diagnosis"),
        ranking_percentile=payload.get("ranking_percentile"),
        hint_level_suggestion=payload.get("hint_level_suggestion"),
    )


@router.get(
    "/students/{user_id}/profile",
    response_model=SkillProfileResponse,
    tags=["Training"]
)
async def get_student_skill_profile(
    user_id: int,
    db: AsyncSession = Depends(get_db)
):
    """UF-10: 获取学员技能画像"""
    service = SkillProfileService(db)
    profile = await service.get_or_create_profile(user_id)

    return SkillProfileResponse(
        user_id=profile.user_id,
        overall_level=profile.overall_level,
        total_sessions=profile.total_sessions,
        total_duration=profile.total_duration,
        last_trained_at=profile.last_trained_at,
        score_safety=float(profile.score_safety) if profile.score_safety is not None else None,
        score_procedure=float(profile.score_procedure) if profile.score_procedure is not None else None,
        score_precision=float(profile.score_precision) if profile.score_precision is not None else None,
        score_efficiency=float(profile.score_efficiency) if profile.score_efficiency is not None else None,
        score_tools=float(profile.score_tools) if profile.score_tools is not None else None,
        cert_l1_passed=profile.cert_l1_passed,
        cert_l2_passed=profile.cert_l2_passed,
        cert_l3_eligible=profile.cert_l3_eligible,
    )


@router.get(
    "/students/{user_id}/weak-steps",
    response_model=List[WeakStepResponse],
    tags=["Training"]
)
async def get_student_weak_steps(
    user_id: int,
    unresolved_only: bool = False,
    limit: int = 20,
    db: AsyncSession = Depends(get_db)
):
    """UF-10: 获取学员薄弱步骤列表"""
    service = SkillProfileService(db)
    weak_steps = await service.get_weak_steps(
        user_id=user_id,
        unresolved_only=unresolved_only,
        limit=limit,
    )

    return [
        WeakStepResponse(
            step_id=step.step_id,
            sop_id=step.sop_id,
            fail_count=step.fail_count,
            last_failed_at=step.last_failed_at,
            fail_tags=step.fail_tags,
            is_resolved=step.is_resolved,
        )
        for step in weak_steps
    ]
